File: src/v2/protocol.cc
#include <v8.h>
#include <node.h>
#include <node_buffer.h>
#include <stdio.h>

#include "spdy.h"

#define UNWRAP(x) \
    Local<Object> buff = args[x]->ToObject();\
    uint8_t* data = reinterpret_cast<uint8_t*>(Buffer::Data(buff));\
    size_t data_len = Buffer::Length(buff);\
    Local<Object> result = Object::New();

namespace spdy {

using namespace v8;
using namespace node;

// parseSynHead (type, flags, data)
Handle<Value> ParseSynHead(const Arguments& args) {
  HandleScope scope;

  uint16_t type = static_cast<uint16_t>(args[0].As<Number>()->Value());
  uint8_t flags = static_cast<uint8_t>(args[1].As<Number>()->Value());
  UNWRAP(2);

  bool syn_stream = type == 0x01;
  uint32_t id = readUInt32(data) & 0x7fffffff;
  uint32_t associated = syn_stream ? (readUInt32(data + 4) & 0x7fffffff) : 0;
  uint8_t priority = syn_stream ? (data[8] >> 6) : 0;
  bool fin = (flags & 0x01) == 0x01;
  bool unidir = (flags & 0x02) == 0x02;
  uint16_t offset = syn_stream ? 10 : 6;

  result->Set(String::NewSymbol("type"),
              syn_stream ?
                  String::NewSymbol("SYN_STREAM")
                  :
                  String::NewSymbol("SYN_REPLY"));
  result->Set(String::NewSymbol("id"), Number::New(id));
  result->Set(String::NewSymbol("associated"), Number::New(associated));
  result->Set(String::NewSymbol("priority"), Number::New(priority));
  result->Set(String::NewSymbol("fin"), fin ? True() : False());
  result->Set(String::NewSymbol("unidir"), unidir ? True() : False());
  result->Set(String::NewSymbol("_offset"), Number::New(offset));

  return scope.Close(result);
}


Handle<Value> ParseHeaders(const Arguments& args) {
  HandleScope scope;

  UNWRAP(0);

  uint16_t size = readUInt16(data);
  // Prevent oob errors
  data += 2;
  data_len -= 2;

  while (size > 0) {
    uint16_t size;
    if (data_len <= 0) return scope.Close(result);

    // Key
    size = readUInt16(data);

    data += 2;
    data_len -= 2;
    if (data_len < size) return scope.Close(result);

    Local<String> key = String::New(reinterpret_cast<char*>(data), size);
    data += size;
    data_len -= size;

    // Value
    size = readUInt16(data);

    data += 2;
    data_len -= 2;
    if (data_len < size) return scope.Close(result);

    Local<String> value = String::New(reinterpret_cast<char*>(data), size);
    data += size;
    data_len -= size;

    result->Set(key, value);
  }

  return scope.Close(result);
}


Handle<Value> ParseRst(const Arguments& args) {
  HandleScope scope;

  UNWRAP(0);

  uint32_t id = readUInt32(data) & 0x7fffffff;
  uint32_t status = readUInt32(data + 4);

  result->Set(String::NewSymbol("type"), String::NewSymbol("RST_STREAM"));
  result->Set(String::NewSymbol("id"), Number::New(id));
  result->Set(String::NewSymbol("status"), Number::New(status));

  return scope.Close(result);
}


Handle<Value> ParseGoaway(const Arguments& args) {
  HandleScope scope;

  UNWRAP(0);

  uint32_t last_id = readUInt32(data) & 0x7fffffff;

  result->Set(String::NewSymbol("type"), String::NewSymbol("GOAWAY"));
  result->Set(String::NewSymbol("lastId"), Number::New(last_id));

  return scope.Close(result);
}


void Initialize(Handle<Object> target) {
  NODE_SET_METHOD(target, "parseSynHead", ParseSynHead);
  NODE_SET_METHOD(target, "parseHeaders", ParseHeaders);
  NODE_SET_METHOD(target, "parseRst", ParseRst);
  NODE_SET_METHOD(target, "parseGoaway", ParseGoaway);
}

NODE_MODULE(protocol, Initialize)

} // spdy

File: lib/spdy/protocol/v2/protocol.js
var protocol = exports;

//
// ### function parseSynHead (type, flags, data)
// #### @type {Number} Frame type
// #### @flags {Number} Frame flags
// #### @data {Buffer} input data
// Returns parsed syn_* frame's head
//
protocol.parseSynHead = function parseSynHead(type, flags, data, callback) {
  var stream = type === 0x01;

  if (data.length < (stream ? 10 : 6)) {
    return callback(new Error('SynHead OOB'));
  }

  callback(null, {
    type: stream ? 'SYN_STREAM' : 'SYN_REPLY',
    id: data.readUInt32BE(0, true) & 0x7fffffff,
    version: 2,
    associated: stream ? data.readUInt32BE(4, true) & 0x7fffffff : 0,
    priority: stream ? data[8] >> 6 : 0,
    fin: (flags & 0x01) === 0x01,
    unidir: (flags & 0x02) === 0x02,
    _offset: stream ? 10 : 6
  });
};

//
// ### function parseHeaders (pairs)
// #### @pairs {Buffer} header pairs
// Returns hashmap of parsed headers
//
protocol.parseHeaders = function parseHeaders(pairs, callback) {
  var count = pairs.readUInt16BE(0, true),
      headers = {};

  pairs = pairs.slice(2);

  function readString() {
    if (pairs.length < 2) {
      return null;
    }
    var len = pairs.readUInt16BE(0, true);

    if (pairs.length < 2 + len) {
      return null;
    }
    var value = pairs.slice(2, 2 + len);

    pairs = pairs.slice(2 + len);

    return value.toString();
  }

  while(count > 0) {
    var key = readString(),
        value = readString();
    if (key === null || value === null) {
      return callback(new Error('Headers OOB'));
    }
    headers[key] = value;
    count--;
  }

  callback(null, headers);
};

//
// ### function parsesRst frame
protocol.parseRst = function parseRst(data, callback) {
  if (data.length < 8) return callback(new Error('RST OOB'));

  callback(null, {
    type: 'RST_STREAM',
    id: data.readUInt32BE(0, true) & 0x7fffffff,
    status: data.readUInt32BE(4, true)
  });
};

protocol.parseGoaway = function parseGoaway(data, callback) {
  if (data.length < 4) return callback(new Error('GOAWAY OOB'));

  callback(null, {
    type: 'GOAWAY',
    lastId: data.readUInt32BE(0, true) & 0x7fffffff
  });
};
